mllib/src/main/scala/org/apache/spark/mllib/nlp/CRF.scala

```diff
@@ -0,0 +1,265 @@
+          threadPool(0).zeroOne += threadPool(i).zeroOne
+        }
+        while (k < featureIndex.maxid) {
+          if (i > 0) {
+            threadPool(0).expected(k) += threadPool(i).expected(k)
+          }
+          threadPool(0).obj += alpha(k) * alpha(k) / 2.0 * C
+          threadPool(0).expected(k) += alpha(k) / C
+          k += 1
+        }
+        k = 0
+        i += 1
+      }
+      i = 0
+      if (itr == 0) {
+        diff = 1.0
+      } else {
+        diff = math.abs((old_obj - threadPool(0).obj) / old_obj)
+      }
+      old_obj = threadPool(0).obj
+      printf("iter=%d, terr=%2.5f, serr=%2.5f, act=%d, obj=%2.5f,diff=%2.5f\n",
+        itr, 1.0 * threadPool(0).err / all,
+        1.0 * threadPool(0).zeroOne / tagger.size, featureIndex.maxid,
+        threadPool(0).obj, diff)
+      if (diff < eta) {
+        converge += 1
+      } else {
+        converge = 0
+      }
+      if (converge == 3) {
+        itr = maxiter + 1 // break
+      }
+      if (diff == 0) {
+        itr = maxiter + 1 // break
+      }
+      opt.optimizer(featureIndex.maxid, alpha, threadPool(0).obj, threadPool(0).expected, C)
+      itr += 1
+    }
+  }
+
+  /**
+   * Use multiple threads to parse the segments
+   * in a unit sentence or paragraph.
+   */
+  class CRFThread extends Thread {
+    var x: ArrayBuffer[Tagger] = null
+    var start_i: Int = 0
+    var err: Int = 0
+    var zeroOne: Int = 0
+    var size: Int = 0
+    var obj: Double = 0.0
+    var expected: ArrayBuffer[Double] = new ArrayBuffer[Double]()
+
+    def initExpected(): Unit = {
+      var i: Int = 0
+      while (i < featureIdx.maxid) {
+        expected.append(0.0)
+        i += 1
+      }
+    }
+
+    /**
+     * Train CRF model and calculate the expectations
+     */
+    override def run(): Unit = {
+      var idx: Int = 0
+      initExpected()
+      while (idx >= start_i && idx < size) {
+        obj += x(idx).gradient(expected)
+        err += x(idx).eval()
+        if (err != 0) {
+          zeroOne += 1
+        }
+        idx = idx + threadNum
+      }
+    }
+  }
+}
+
+@DeveloperApi
+object CRF {
+  @transient var sc: SparkContext = _
+
+  /**
+   * Train CRF Model
+   * A word's context is its sentence or its nearby
+   * paragraph and the sentence or nearby paragraph will
+   * not be very long. So A sentence or paragraph will be
+   * processed in a node to reduce partition and networking
+   * costs. Multiple sentences or paragraphs are
+   * collected from multiple nodes to create the overall result.
+   *
+   * Feature file format
+   * word|word characteristic|designated label
+   *
+   * @param templates Source templates for training the model
+   * @param features Source files for training the model
+   * @return Model of a unit
+   */
+  def runCRF(templates: RDD[Array[String]],
+             features: RDD[Array[String]]): CRFModel = {
+    val crf = new CRF()
+    val template: Array[Array[String]] = templates.toLocalIterator.toArray
```

Comment:
This is not running in a Spark cluster model, this will be run only at the local mode.